File: lint-arwaky/archive/capabilities/scope_boundary_analyzer.py
"""scope_boundary_analyzer — JS/TS scope boundary detection."""

from __future__ import annotations
from ..taxonomy import FileContentVO, FilePath, LineContentVO, LineNumber, SymbolName

import re

from ..contract import IScopeBoundaryProtocol
from ..contract import IFileSystemPort


_FUNCTION_PATTERNS: list[re.Pattern] = [
    re.compile(r"(?:async\s+)?function\s+([A-Za-z_$][A-Za-z0-9_$]*)\s*\("),
    re.compile(
        r"(?:const|let|var)\s+([A-Za-z_$][A-Za-z0-9_$]*)\s*=\s*(?:async\s*)?(?:\([^)]*\)|[A-Za-z_$][A-Za-z0-9_$]*)\s*=>"
    ),
    re.compile(
        r"^(?:async\s+|static\s+|private\s+|protected\s+|public\s+)*([A-Za-z_$][A-Za-z0-9_$]*)\s*\("
    ),
]

_CLASS_PATTERN = re.compile(
    r"class\s+([A-Za-z_$][A-Za-z0-9_$]*)(?:\s+extends\s+[A-Za-z_$][A-Za-z0-9_$]*)?"
)


class ScopeBoundaryAnalyzer(IScopeBoundaryProtocol):
    """Capability for detecting JS/TS scope boundaries."""

    def __init__(self, fs: IFileSystemPort):
        self.fs = fs

    def detect_js_scope(self, stripped_line: LineContentVO) -> SymbolName | None:
        res = self._detect_js_scope_raw(str(stripped_line))
        return SymbolName(value=res) if res else None

    def find_scope_bounds(
        self, content: FileContentVO, scope_line: LineNumber | None
    ) -> tuple[LineNumber | None, LineNumber | None]:
        lines = str(content).splitlines()
        start, end = self._find_scope_bounds_raw(
            lines, int(scope_line) if scope_line is not None else None
        )
        return (
            LineNumber(value=start) if start is not None else None,
            LineNumber(value=end) if end is not None else None,
        )

    def get_enclosing_scope(
        self, file_path: FilePath, line: LineNumber
    ) -> SymbolName | None:
        res = self._get_enclosing_scope_raw(file_path, line)
        return SymbolName(value=res) if res else None

    def _detect_js_scope_raw(self, stripped_line: str) -> str | None:
        """Detect if a stripped JS/TS line opens a named scope."""
        match = _CLASS_PATTERN.search(stripped_line)
        if match:
            return f"class {match.group(1)}"
        for pattern in _FUNCTION_PATTERNS:
            match = pattern.search(stripped_line)
            if match:
                name = match.group(1)
                if name not in {"if", "for", "while", "switch", "catch", "else"}:
                    return f"function {name}"
        return None

    def _find_scope_bounds_raw(
        self, lines: list[str], scope_line: int | None
    ) -> tuple[int | None, int | None]:
        """Find start/end line numbers of enclosing function body via brace counting."""
        if scope_line is None:
            return None, None
        brace_depth = 0
        scope_start: int | None = None
        scope_end: int | None = None
        for i in range(scope_line - 1, len(lines)):
            line = lines[i]
            if "{" in line and scope_start is None:
                scope_start = i + 1
                brace_depth = 1
                continue
            if scope_start is not None:
                brace_depth += line.count("{") - line.count("}")
                if brace_depth <= 0:
                    scope_end = i + 1
                    break
        return scope_start, scope_end

    def _update_scope_stack(
        self,
        stack: list[tuple[str, int]],
        depth: int,
        detected_scope: str | None,
        raw_line: str,
    ) -> tuple[list[tuple[str, int]], int]:
        """Pop expired scopes, optionally push new, then update depth and pop again."""
        while stack and depth <= stack[-1][1]:
            stack.pop()
        if detected_scope and "{" in raw_line:
            stack.append((detected_scope, depth))
        depth += raw_line.count("{") - raw_line.count("}")
        while stack and depth <= stack[-1][1]:
            stack.pop()
        return stack, depth

    def _get_enclosing_scope_raw(
        self, file_path: FilePath, line: LineNumber
    ) -> str | None:
        """Find the nearest enclosing function or class for a given 1-indexed line."""
        if not self.fs.is_file(file_path).value:
            return None

        try:
            content = self.fs.read_file(file_path).value
            lines = content.splitlines()
        except Exception:
            return None

        scope_stack: list[tuple[str, int]] = []
        brace_depth = 0

        for i, raw_line in enumerate(lines):
            current_line_no = i + 1
            stripped = raw_line.strip()
            detected_scope = self._detect_js_scope_raw(stripped)

            scope_stack, brace_depth = self._update_scope_stack(
                scope_stack, brace_depth, detected_scope, raw_line
            )

            if current_line_no == int(line):
                break

        return " -> ".join(s[0] for s in scope_stack) if scope_stack else None

File: lint-arwaky/src-rust/capabilities/scope_boundary_analyzer.rs
// scope_boundary_analyzer — JS/TS scope boundary detection.
// Implements IScopeBoundaryProtocol: detect_js_scope, find_scope_bounds, get_enclosing_scope.

use std::fs;
use regex::Regex;
use once_cell::sync::Lazy;
use crate::taxonomy::{FilePath, LineNumber};

/// Patterns to detect JS/TS function definitions
static FUNCTION_PATTERNS: Lazy<Vec<Regex>> = Lazy::new(|| vec![
    // Standard function declarations
    Regex::new(r"(?:async\s+)?function\s+([A-Za-z_$][A-Za-z0-9_$]*)\s*\(").unwrap(),
    // Arrow functions assigned to const/let/var
    Regex::new(r"(?:const|let|var)\s+([A-Za-z_$][A-Za-z0-9_$]*)\s*=\s*(?:async\s*)?(?:\([^)]*\)|[A-Za-z_$][A-Za-z0-9_$]*)\s*=>").unwrap(),
    // Method definitions in classes
    Regex::new(r"^\s*(?:async\s+|static\s+|private\s+|protected\s+|public\s+)*([A-Za-z_$][A-Za-z0-9_$]*)\s*\(").unwrap(),
]);

static CLASS_PATTERN: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"class\s+([A-Za-z_$][A-Za-z0-9_$]*)(?:\s+extends\s+[A-Za-z_$][A-Za-z0-9_$]*)?").unwrap()
});

const JS_KEYWORDS: &[&str] = &["if", "for", "while", "switch", "catch", "else"];

/// Capability for detecting JS/TS scope boundaries.
pub struct ScopeBoundaryAnalyzer;

impl ScopeBoundaryAnalyzer {
    pub fn new() -> Self {
        Self
    }

    /// Detect if a stripped JS/TS line opens a named scope.
    pub fn detect_js_scope(&self, stripped_line: &str) -> Option<String> {
        // Check for class definition first
        if let Some(captures) = CLASS_PATTERN.captures(stripped_line) {
            if let Some(name) = captures.get(1) {
                return Some(format!("class {}", name.as_str()));
            }
        }

        // Check for function patterns
        for pattern in FUNCTION_PATTERNS.iter() {
            if let Some(captures) = pattern.captures(stripped_line) {
                if let Some(name) = captures.get(1) {
                    let n = name.as_str();
                    if !JS_KEYWORDS.contains(&n) {
                        return Some(format!("function {}", n));
                    }
                }
            }
        }

        None
    }

    /// Find start/end line numbers of enclosing function body via brace counting.
    pub fn find_scope_bounds(
        &self,
        content: &str,
        scope_line: Option<usize>,
    ) -> (Option<usize>, Option<usize>) {
        let scope_line = match scope_line {
            Some(l) => l,
            None => return (None, None),
        };

        let lines: Vec<&str> = content.lines().collect();
        let mut brace_depth: i32 = 0;
        let mut scope_start: Option<usize> = None;
        let mut scope_end: Option<usize> = None;

        for i in (scope_line.saturating_sub(1))..lines.len() {
            let line = lines[i];
            if line.contains('{') && scope_start.is_none() {
                scope_start = Some(i + 1);
                brace_depth = 1;
                continue;
            }
            if scope_start.is_some() {
                brace_depth += line.chars().filter(|&c| c == '{').count() as i32;
                brace_depth -= line.chars().filter(|&c| c == '}').count() as i32;
                if brace_depth <= 0 {
                    scope_end = Some(i + 1);
                    break;
                }
            }
        }

        (scope_start, scope_end)
    }

    /// Find the nearest enclosing function or class for a given 1-indexed line.
    pub fn get_enclosing_scope(&self, file_path: &str, target_line: usize) -> Option<String> {
        let content = fs::read_to_string(file_path).ok()?;
        let lines: Vec<&str> = content.lines().collect();

        // Stack of (scope_name, brace_depth_when_opened)
        let mut scope_stack: Vec<(String, i32)> = Vec::new();
        let mut brace_depth: i32 = 0;

        for (i, raw_line) in lines.iter().enumerate() {
            let current_line_no = i + 1;
            let stripped = raw_line.trim();

            // Pop expired scopes
            while let Some((_, depth)) = scope_stack.last() {
                if brace_depth <= *depth {
                    scope_stack.pop();
                } else {
                    break;
                }
            }

            // Detect and push new scope
            if let Some(scope_name) = self.detect_js_scope(stripped) {
                if raw_line.contains('{') {
                    scope_stack.push((scope_name, brace_depth));
                }
            }

            // Apply brace count for this line
            let open = raw_line.chars().filter(|&c| c == '{').count() as i32;
            let close = raw_line.chars().filter(|&c| c == '}').count() as i32;
            brace_depth += open - close;

            // Pop scopes that closed on this line
            while let Some((_, depth)) = scope_stack.last() {
                if brace_depth <= *depth {
                    scope_stack.pop();
                } else {
                    break;
                }
            }

            if current_line_no == target_line {
                break;
            }
        }

        if scope_stack.is_empty() {
            None
        } else {
            Some(scope_stack.iter().map(|(name, _)| name.as_str()).collect::<Vec<_>>().join(" -> "))
        }
    }
}
